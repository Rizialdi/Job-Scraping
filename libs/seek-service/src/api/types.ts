/**
 * Base configuration for the SeekService
 */
interface ISeekServiceBaseConfig {
  baseUrl: string;
  api: {
    v4: {
      baseUrl: string;
      endpoints: {
        search: string;
        relatedSearch: string;
      };
    };
  };
}

/**
 * Job listing from the search endpoint
 */
interface ISeekServiceSearchResponseJobListing {
  advertiser: {
    id: string;
    description: string;
  };
  area?: string;
  areaId?: number;
  areaWhereValue?: string;
  automaticInclusion: boolean;
  branding?: {
    id: string;
    assets: {
      logo: {
        strategies: {
          jdpLogo: string;
          serpLogo: string;
        };
      };
    };
  };
  bulletPoints: Array<string>;
  classification: {
    id: string;
    description: string;
  };
  companyName?: string;
  companyProfileStructuredDataId: number;
  displayType: string;
  listingDateDisplay: string;
  location: string;
  locationId: number;
  locationWhereValue: string;
  id: number;
  isPremium: boolean;
  isStandOut: boolean;
  jobLocation: {
    label: string;
    countryCode: string;
    seoHierarchy: {
      contextualName: string;
    }[];
  };
  listingDate: string;
  logo: {
    id: string;
    description: unknown;
  };
  roleId: string;
  salary: string;
  solMetadata: {
    searchRequestToken: string;
    token: string;
    jobId: string;
    section: string;
    sectionRank: number;
    jobAdType: string;
    tags: {
      mordor__flights: string;
      mordor__s: string;
    };
  };
  subClassification: {
    id: string;
    description: string;
  };
  suburb?: string;
  suburbId?: number;
  suburbWhereValue?: string;
  teaser: string;
  title: string;
  tracking: string;
  workType: string;
  isPrivateAdvertiser: boolean;
  displayStyle?: {
    search?: string;
  };
}

/**
 * Response from the search endpoint
 */
interface ISeekServiceSearchResponse {
  data: ISeekServiceSearchResponseJobListing[];
  title: string;
  totalCount: number;
  paginationParameters: {
    seekSelectAllPages: boolean;
    hadPremiumListings: boolean;
  };
  info: {
    timeTaken: number;
    source: string;
    experiment: string;
  };
  userQueryId: string;
  sortMode: {
    isActive: boolean;
    name: string;
    value: string;
  }[];
  solMetadata: {
    requestToken: string;
    token: string;
    keywords: string;
    sortMode: string;
    pageSize: number;
    pageNumber: number;
    totalJobCount: number;
    tags: {
      mordor__flights: string;
      'chalice-search-api:solId': string;
    };
  };
  joraCrossLink: {
    canCrossLink: boolean;
  };
  searchParams: {
    page: string;
    keywords: string;
    solid: string;
  };
}

export {
  ISeekServiceBaseConfig,
  ISeekServiceSearchResponse,
  ISeekServiceSearchResponseJobListing,
};
