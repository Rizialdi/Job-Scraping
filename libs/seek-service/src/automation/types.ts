/**
 * Parse the data-search-sol-meta attribute from the job ad
 */
interface IDataSearchSolMeta {
  searchRequestToken: '333bc538-79f2-4e09-b6f2-60afa8ebee57';
  token: '0~333bc538-79f2-4e09-b6f2-60afa8ebee57';
  jobId: '59820105';
  section: 'MAIN';
  sectionRank: 22;
  jobAdType: 'ORGANIC';
  tags: {
    mordor__flights: 'mordor_123';
    mordor__s: '0';
    'seek:chalice:experiments:job_details_split_view': '2';
    'seek:chalice:experiments:aa_test': '2';
  };
}

/**
 * Job listing interface from scraping the Seek website
 */
interface IJobListing {
  companyName: string;
  jobUrl: string;
  jobId: string;
  location: string;
  jobTitle: string;
  jobListingDate: string;
  jobClassification: string;
  jobSpecifics: string;
  jobSalaryRange: string;
  jobContractType: string;
  jobSalaryType: string;
  jobAdType: string;
  roleId: string;
  logo: string;
  jobTeaser: string;
}

export { IDataSearchSolMeta, IJobListing };
