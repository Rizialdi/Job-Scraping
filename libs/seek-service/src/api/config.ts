import { ISeekServiceBaseConfig } from './types';

/**
 * Default configuration for the SeekService
 */
const defaultConfiguration: ISeekServiceBaseConfig = {
  baseUrl: 'https://www.seek.com.au',
  api: {
    v4: {
      baseUrl: '/api/chalice-search/v4',
      endpoints: {
        search: '/search?page={pageNumber}&keywords={keyword}',
        relatedSearch:
          '/related-search?zone=anz-1&siteKey=au&keywords={keyword}',
      },
    },
  },
};

export { defaultConfiguration };
