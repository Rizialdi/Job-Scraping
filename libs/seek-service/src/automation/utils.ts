import puppeteer from 'puppeteer';
import { IDataSearchSolMeta, IJobListing } from './types';

/**
 * Search for a keyword on Seek website
 * @param keyword - The keyword to search for
 * @returns The list of job listings
 */
const search = async (keyword: string): Promise<IJobListing[]> => {
  const browser = await puppeteer.launch();
  const page = await browser.newPage();

  const url = 'https://www.seek.com.au';

  // Navigate to the website
  await page.goto(url);

  // Set screen size
  await page.setViewport({ width: 1080, height: 1024 });

  // Find the search box
  const keywordAnchor = await page.$("input[placeholder='Enter Keywords']");
  // Find the search button
  const searchAnchor = await page.$("button[id='searchButton']");

  if (!keywordAnchor) throw new Error('No anchor found');
  if (!searchAnchor) throw new Error('No anchor found');

  // Type into search box
  await keywordAnchor.type(keyword);
  // Click on search button
  await searchAnchor.click();

  // Wait for the results to load
  await page.waitForSelector('div._1wkzzau0.a1msqi6j > div._1wkzzau0');

  // Iterate over the results and extract the data
  const jobListings = await page.evaluate((url) => {
    return Array.from(
      document.querySelectorAll('div._1wkzzau0.a1msqi6j > div._1wkzzau0')
    ).map((jobListing) => {
      const meta = jobListing
        .querySelector('div > div')
        ?.getAttribute('data-search-sol-meta');

      const { jobId, jobAdType }: IDataSearchSolMeta = JSON.parse(meta);

      return {
        companyName:
          jobListing.querySelector(
            'div > div > article > div > div > div > div > div > div > div > div > div > span > a'
          )?.textContent ?? 'Private Advertiser',
        jobUrl:
          url + jobListing.querySelector('div > a')?.getAttribute('href') ?? '',
        jobId: jobId,
        location:
          jobListing.querySelector(
            'div > div > article > div > div > div > div > div > div > div > span > a'
          )?.textContent ?? '',
        jobTitle:
          jobListing.querySelector(
            'div > div > article > div > div > div > div > div > div > div > div > div > div > h3 > a'
          )?.textContent ?? '',
        jobListingDate:
          jobListing.querySelector(
            'div > div > article > div > div > div > div > div > div > div > div > span > span'
          )?.textContent ?? '',
        jobClassification:
          jobListing.querySelector(
            'div > div > article > div > div > div > div > div > div > div > span > div > div > div > a'
          )?.textContent ?? '',
        jobSpecifics:
          jobListing.querySelector(
            'div > div > article > div > div > div > div > div > span'
          )?.textContent ?? '',
        jobSalaryRange:
          jobListing.querySelector(
            'div > div > article > div > div > div > div > div > div > div > div > span > div'
          )?.textContent ?? '',
        jobContractType:
          jobListing.querySelector('div > div> div > div > div > div > p')
            ?.textContent ?? '',
        jobSalaryType: '',
        jobAdType: jobAdType,
        roleId: '',
        logo:
          jobListing
            .querySelector('div > div > div > div > div > img')
            ?.getAttribute('src') ?? '',
        jobTeaser:
          jobListing.querySelector(
            'div > div > article > div > div > div > div > div > span'
          )?.textContent ?? '',
      };
    });
  }, url);

  await browser.close();

  return jobListings;
};

export { search };
