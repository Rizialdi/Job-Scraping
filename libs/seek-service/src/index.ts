export * from './lib/seek-service';
export * from './lib/types';
