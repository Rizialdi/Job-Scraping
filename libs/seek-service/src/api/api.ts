import axios, { AxiosInstance } from 'axios';
import { defaultConfiguration } from './config';
import {
  ISeekServiceApiBaseConfig,
  ISeekServiceApiRelatedSearchResponse,
  ISeekServiceApiSearchResponse,
} from './types';

/**
 * SeekService API
 * This class is responsible for making requests to the Seek API
 */
class SeekServiceApi {
  public config: ISeekServiceApiBaseConfig;
  private axiosInstance: AxiosInstance;

  constructor() {
    this.config = defaultConfiguration;

    this.axiosInstance = axios.create({
      baseURL: this.config.baseUrl,
      timeout: 5000,
    });
  }

  /**
   * Get the current configuration for the SeekService
   * @returns The current configuration for the SeekService
   */
  public getConfiguration(): ISeekServiceApiBaseConfig {
    return this.config;
  }

  /**
   * Get job listings given a keyword and a given page number
   * @param keyword - The keyword to search for
   * @param pageNumber - The page number to search for
   * @returns The response from the API
   */
  public async search(
    keyword: string,
    pageNumber: number
  ): Promise<ISeekServiceApiSearchResponse> {
    const {
      api: {
        v4: { baseUrl, endpoints },
      },
    } = this.config;

    const url = `${baseUrl}${endpoints.search}`
      .replace('{pageNumber}', pageNumber.toString())
      .replace('{keyword}', keyword);

    const { data } = await this.axiosInstance.get(url);

    return data;
  }

  /**
   * Get related keywords given a keyword
   * @param keyword - The keyword to search for
   * @returns The response from the API
   */
  public async relatedSearch(
    keyword: string
  ): Promise<ISeekServiceApiRelatedSearchResponse> {
    const {
      api: {
        v4: { baseUrl, endpoints },
      },
    } = this.config;

    const url = `${baseUrl}${endpoints.relatedSearch}`.replace(
      '{keyword}',
      keyword
    );

    const { data } = await this.axiosInstance.get(url);

    return data;
  }
}

export { SeekServiceApi };
