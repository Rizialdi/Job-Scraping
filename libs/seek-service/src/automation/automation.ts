import { IJobListing } from './types';
import { search } from './utils';

/**
 * SeekService Automation
 * This class is responsible for scraping the Seek website
 */
class SeekServiceAutomation {
  /**
   * Get job listings given a keyword
   * @param keyword - The keyword to search for
   * @returns The response from the API
   */
  static async search(keyword: string): Promise<IJobListing[]> {
    return search(keyword);
  }
}

export { SeekServiceAutomation };
