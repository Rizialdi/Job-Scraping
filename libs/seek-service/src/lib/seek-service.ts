import { SeekServiceApi } from '../api/api';
import { ISeekServiceApiSearchResponseJobListing } from '../api/types';
import { SeekServiceAutomation } from '../automation/automation';
import { ISeekServiceSearchResponseJobListing } from './types';
import { dayjs } from './utils';

/**
 * This class is responsible for abstracting the SeekService API and providing the ability to manipulate the data
 */
class SeekService {
  private api: SeekServiceApi;
  private automation: SeekServiceAutomation;

  constructor() {
    this.api = new SeekServiceApi();
    this.automation = new SeekServiceAutomation();
  }

  /**
   * Get the result of a search given a keyword and a given page number
   * @param keyword - The keyword to search for
   * @param pageNumber - The page number to search for
   * @returns The list of found jobs
   */
  public async search(
    keyword: string,
    pageNumber: number
  ): Promise<ISeekServiceApiSearchResponseJobListing[]> {
    const { data } = await this.api.search(keyword, pageNumber);

    return data;
  }

  /**
   * Get the related keyword of a given keyword
   * @param keyword - The keyword to get the related keyword for
   * @returns The list of related keywords
   */
  public async relatedKeywords(keyword: string): Promise<string[]> {
    const { relatedSearches } = await this.api.relatedSearch(keyword);

    const filteredSearch = relatedSearches.filter(
      (relatedSearch) => relatedSearch.type === 'keyword'
    );

    return filteredSearch.map((relatedSearch) => relatedSearch.keywords);
  }

  /**
   * Get the result of a search given a keyword using the browser
   * @param keyword - The keyword to search for
   * @returns The list of found jobs
   */
  public async searchUsingBrowser(
    keyword: string
  ): Promise<ISeekServiceSearchResponseJobListing[]> {
    const response = await this.automation.search(keyword);

    return response.map((jobListing) => ({
      company_name: jobListing.companyName,
      Extracted_at: dayjs.utc().format('ddd, DD MMM YYYY HH:mm:ss [GMT]'),
      seek_job_profile_url: jobListing.jobUrl,
      Job_id: jobListing.jobId,
      Location: jobListing.location,
      job_title: jobListing.jobTitle,
      Job_listing_date: dayjs(jobListing.jobListingDate).format(
        'ddd, DD MMM YYYY'
      ),
      job_classification: jobListing.jobClassification,
      seek_specifics: jobListing.jobSpecifics,
      salary_range: jobListing.jobSalaryRange,
      contract_type: jobListing.jobContractType
        .split('This is a')[1]
        .split('job')[0]
        .trim(),
      salary_type: jobListing.jobSalaryType,
      seek_job_ad_type: jobListing.jobAdType,
      seek_role_id: jobListing.roleId,
      logo: jobListing.logo,
      seek_job_teaser: jobListing.jobTeaser,
    }));
  }
}

export { SeekService };
