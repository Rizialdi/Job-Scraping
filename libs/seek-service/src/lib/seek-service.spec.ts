import { relatedSearch, searchResult } from '../__tests__/stubs/api.stub';
import { searchResult as searchResultAutomation } from '../__tests__/stubs/automation.stub';
import { SeekServiceApi } from '../api/api';
import { SeekServiceAutomation } from '../automation/automation';
import { SeekService } from './seek-service';

describe('#SeekService', () => {
  let sut: SeekService;
  let keyword: string;
  let pageNumber: number;
  let searchMock: jest.SpyInstance;
  let searchUsingBrowserMock: jest.SpyInstance;
  let relatedSearchMock: jest.SpyInstance;

  beforeEach(() => {
    sut = new SeekService();

    pageNumber = 1;
    keyword = 'fullstack';

    searchMock = jest.spyOn(SeekServiceApi.prototype, 'search');
    searchUsingBrowserMock = jest.spyOn(
      SeekServiceAutomation.prototype,
      'search'
    );
    relatedSearchMock = jest.spyOn(SeekServiceApi.prototype, 'relatedSearch');
  });

  afterEach(() => {
    searchMock.mockRestore();
    searchUsingBrowserMock.mockRestore();
    relatedSearchMock.mockRestore();
  });

  describe('#constructor', () => {
    it('should be defined', () => {
      expect(sut).toBeDefined();
    });
  });

  describe('#search', () => {
    it('should call the correct method', async () => {
      await sut.search(keyword, pageNumber);

      expect(searchMock).toHaveBeenCalledWith(keyword, pageNumber);
    });

    it('should have the response matching the snapshot', async () => {
      searchMock.mockResolvedValue(searchResult());

      const response = await sut.search(keyword, pageNumber);

      expect(response).toMatchSnapshot();
    });
  });

  describe('#relatedKeywords', () => {
    it('should call the correct method', async () => {
      await sut.relatedKeywords(keyword);

      expect(relatedSearchMock).toHaveBeenCalledWith(keyword);
    });

    it('should have the response matching the snapshot', async () => {
      relatedSearchMock.mockResolvedValue(relatedSearch());

      const response = await sut.relatedKeywords(keyword);

      expect(response).toMatchSnapshot();
    });
  });

  describe('#searchUsingBrowser', () => {
    it('should call the correct method', async () => {
      await sut.searchUsingBrowser(keyword);

      expect(searchUsingBrowserMock).toHaveBeenCalledWith(keyword);
    });

    it('should have the response matching the snapshot', async () => {
      searchUsingBrowserMock.mockResolvedValue(searchResultAutomation());

      const response = await sut.searchUsingBrowser(keyword);

      expect(response).toMatchSnapshot();
    });
  });
});
