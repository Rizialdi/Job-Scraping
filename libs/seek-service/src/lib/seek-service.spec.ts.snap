// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#SeekService #relatedKeywords should have the response matching the snapshot 1`] = `
[
  "java",
  "angular",
  "react",
  "full stack",
  "python",
  "nodejs",
  "javascript",
]
`;

exports[`#SeekService #search should have the response matching the snapshot 1`] = `
[
  {
    "advertiser": {
      "description": "Bridge Associates APAC Pty Ltd",
      "id": "28815495",
    },
    "area": "North Shore & Northern Beaches",
    "areaId": 5028,
    "areaWhereValue": "North Shore & Northern Beaches Sydney NSW",
    "automaticInclusion": false,
    "bulletPoints": [
      "Exciting New Role -",
      "Career Development",
    ],
    "classification": {
      "description": "Information & Communication Technology",
      "id": "6281",
    },
    "companyName": "Bridge Associates APAC",
    "companyProfileStructuredDataId": 221565,
    "displayStyle": {
      "search": "A",
    },
    "displayType": "promoted",
    "id": 61986871,
    "isPremium": true,
    "isPrivateAdvertiser": false,
    "isStandOut": true,
    "jobLocation": {
      "countryCode": "AU",
      "label": "North Sydney, Sydney NSW",
      "seoHierarchy": [
        {
          "contextualName": "North Sydney NSW 2060",
        },
        {
          "contextualName": "All Sydney NSW",
        },
      ],
    },
    "listingDate": "2023-02-28T12:03:01Z",
    "listingDateDisplay": "4d ago",
    "location": "Sydney",
    "locationId": 1000,
    "locationWhereValue": "All Sydney NSW",
    "logo": {
      "description": null,
      "id": "",
    },
    "roleId": "digital-lead",
    "salary": "$140,000 - $169,999",
    "solMetadata": {
      "jobAdType": "SPONSORED",
      "jobId": "61986871",
      "searchRequestToken": "840c4c61-9bf7-4a82-b8f3-cddea9a8f793",
      "section": "MAIN",
      "sectionRank": 1,
      "tags": {
        "mordor__flights": "mordor_123",
        "mordor__s": "0",
      },
      "token": "0~840c4c61-9bf7-4a82-b8f3-cddea9a8f793",
    },
    "subClassification": {
      "description": "Developers/Programmers",
      "id": "6287",
    },
    "suburb": "North Sydney",
    "suburbId": 24000,
    "suburbWhereValue": "North Sydney NSW 2060",
    "teaser": "Exciting new digital lead developer role working with the latest Microsoft techs!!",
    "title": "Digital Lead - Hands on Fullstack Lead  - Agile - DevOps -",
    "tracking": "ewogICJ0b2tlbiI6ICI1OTYyZjM4NS1kNDM0LTk2NGUtOTM1ZC00ZWMwZTMwZTBiNmZfMSIKfQ==",
    "workType": "Full Time",
  },
]
`;

exports[`#SeekService #searchUsingBrowser should have the response matching the snapshot 1`] = `
[
  {
    "Extracted_at": "Mon, 06 Mar 2023 08:00:00 GMT",
    "Job_id": "59820105",
    "Job_listing_date": "Sat, 04 Feb 2023",
    "Location": "Sydney NSW",
    "company_name": "Evolution Recruitment Solutions Pty Ltd",
    "contract_type": "Full Time",
    "job_classification": "Information & Communication Technology",
    "job_title": "Senior  Fullstack .NET Software Engineer",
    "logo": "",
    "salary_range": "$140,000 - $169,999",
    "salary_type": "",
    "seek_job_ad_type": "ORGANIC",
    "seek_job_profile_url": "https://www.seek.com.au/job/59820105?type=standout#sol=09b8a51e2d80eebd75c222500c3ad8ff7159f322",
    "seek_job_teaser": "Your team will be responsible building the core product and be the Engine behind the success of Unicorn Tech Disruptor.",
    "seek_role_id": "",
    "seek_specifics": "Your team will be responsible building the core product and be the Engine behind the success of Unicorn Tech Disruptor.",
  },
]
`;
